### PR TITLE
fix(@angular-ru/websocket): close socket connection on error & serializers

### PR DESCRIPTION
**Pull request by Schirbak**
...and provide serialize & deserialize methods to override

## PR Checklist

Please check if your PR fulfills the following requirements:

-   [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
-   [x] Tests for the changes have been added (for bug fixes / features)
-   [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/websocket/src/abstract-websocket.client.ts [hunk 1]
```diff
@@ -96,26 +100,30 @@ export abstract class AbstractWebsocketClient<K> implements WebsocketHandler<K>,
         this.disconnect();
     }
 
+    protected serialize(message: Any): WebSocketMessage {
+        return JSON.stringify(message);
+    }
+
+    protected deserialize(messageEvent: MessageEvent): WebsocketMessage<K, Any> {
+        return JSON.parse(messageEvent.data);
```
Comment (splincode) [inline packages/websocket/src/abstract-websocket.client.ts:108]: What do you think about?

image

File: packages/websocket/src/abstract-websocket.client.ts [hunk 2]
```diff
@@ -96,26 +100,30 @@ export abstract class AbstractWebsocketClient<K> implements WebsocketHandler<K>,
...
-            });
+            },
+            (): void => this.reconnect()
+        );
     }
 
-    private reconnection(subscription: Subscription): void {
-        if (!subscription.closed) {
-            subscription.unsubscribe();
+    private reconnect(): void {
```
Comment (splincode) [inline packages/websocket/src/abstract-websocket.client.ts:121]: It will be cool if we add tests for this package

**Review by splincode — CHANGES_REQUESTED** [review]

File: packages/websocket/src/abstract-websocket.client.ts (hunk 1 above)
Comment (Schirbak) [reply packages/websocket/src/abstract-websocket.client.ts:108]: I can't imagine a case when `undefined` will get there. But even if your connection can return this, you can override these methods.

File: packages/websocket/src/abstract-websocket.client.ts (hunk 2 above)
Comment (Schirbak) [reply packages/websocket/src/abstract-websocket.client.ts:121]: Ladno

File: packages/websocket/src/abstract-websocket.client.ts (hunk 1 above)
Comment (splincode) [reply packages/websocket/src/abstract-websocket.client.ts:108]: case

```
// server send `ping` instead `{ data: null, result: null }`

// your code
// when messageEvent is "ping" -> "ping".data =>> undefined
protected deserialize(messageEvent: MessageEvent): WebsocketMessage<K, Any> {
        return JSON.parse(messageEvent.data);
}
```

Comment (splincode) [reply packages/websocket/src/abstract-websocket.client.ts:108]: perhaps we should abandon the concept of `MessageEvent`, what do you think?

Comment (Schirbak) [reply packages/websocket/src/abstract-websocket.client.ts:108]: MessageEvent is a JS Event, it carries a lot of metadata among the data. if the message is `ping`, the `messageEvent` will be `{ ..., data: 'ping' }`. 
i guess you mean `WebsocketMessage<K, T>`. i agree, `WebsocketMessage<K, T>` is something weird. but abandon involves a bunch of fixes, no need for this at this moment it think.

MessageEvent is a JS event, it carries a lot of metadata among `data`. if the message is "ping", `messageEvent` will be `{ ..., data: 'ping' }`.
I think you mean `WebsocketMessage <K, T>`. I agree, `WebsocketMessage <K, T>` is weird. but the refusal suggests a bunch of fixes, this is not necessary at the moment I think.

File: packages/websocket/src/abstract-websocket.client.ts [hunk 3]
```diff
@@ -34,8 +35,11 @@ export abstract class AbstractWebsocketClient<K> implements WebsocketHandler<K>,
         return `${this.baseUrl}${this.handlerPath}`;
     }
 
+    // eslint-disable-next-line max-lines-per-function
```
Comment (splincode) [inline packages/websocket/src/abstract-websocket.client.ts:38]: can we get rid of this line of code?

File: packages/websocket/src/abstract-websocket.client.ts (hunk 1 above)
Comment (splincode) [reply packages/websocket/src/abstract-websocket.client.ts:108]: okey

File: packages/websocket/src/abstract-websocket.client.ts [hunk 4]
```diff
@@ -96,26 +87,42 @@ export abstract class AbstractWebsocketClient<K> implements WebsocketHandler<K>,
...
-                        this.reconnection(this.subscription);
-                        return throwError(err);
-                    }
-                ),
-                takeUntil(this.destroy$)
-            )
-            .subscribe((message: WebsocketMessage<K, Any>): void => {
+        this.subscription = this.socket$.pipe(takeUntil(this.destroy$)).subscribe(
+            (message: WebsocketMessage<K, Any>): void => {
                 window.requestAnimationFrame((): void => this.messages$.next(message));
```
Comment (splincode) [inline packages/websocket/src/abstract-websocket.client.ts:114]: What do you think about remove `requestAnimationFrame`?) 

can we somehow correctly throttling the event? so as not to interfere with the events of change detection

**Review by evigkein — APPROVED** [review]

**Review by splincode — APPROVED** [review]

**Pull request merged**